scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala

```diff
@@ -363,6 +365,93 @@ class FormatWriter(formatOps: FormatOps) {
       finalResult.result()
     }
   }
+
+  import scala.meta.internal.classifiers.classifier
+
+  @classifier
+  trait CloseDelim
+  object CloseDelim {
+    def unapply(token: Token): Boolean =
+      token.is[RightParen] || token.is[RightBracket] || token.is[RightBrace]
+  }
+
+  private def handleTrailingCommasAndWhitespace(
+      formatToken: FormatToken,
+      state: State,
+      sb: StringBuilder,
+      whitespace: String): Unit = {
+
+    import org.scalafmt.config.TrailingCommas
+
+    if (!runner.dialect.allowTrailingCommas) {
+      sb.append(whitespace)
+      return
+    }
+
+    val left = formatToken.left
+    val right = nextNonComment(formatToken).right
+    val isNewline = state.splits.last.modification.isNewline
+
+    initStyle.trailingCommas match {
+      // foo(
+      //   a,
+      //   b
+      // )
+      //
+      // Insert a comma after b
+      case TrailingCommas.always
+          if !left.is[Comma] && !left.is[Comment] &&
+            right.is[CloseDelim] && isNewline =>
+        sb.append(",")
```

Comment:
Does this force a comma in this case?

```scala
// before
lst.map { x =>
  x + 1
}
// after
lst.map { x =>
  x + 1,
}
```

I think we may have to limit this behavior to a couple of nodes

- Term.Apply
- Importee/Importer
- Type.Apply

Reply:
Right, tuples as well 👍 

`FormatOps.owners(Token): Tree` goes to the closest enclosing node of a token

Reply:
done! The final list includes also `Defn.Def` and `Decl.Def`. I've added test cases for all of those.